File: cpp-tiny-virtual-machine/src/bytecode.cpp
#include "bytecode.h"
#include "iostream"
void emit1(Script *script, OP_TYPE type);
void emit2(Script *script, OP_TYPE type, uint8_t op1);
void generateCode(Script *script, ASTNode *node, Scope *slink, string indent);
void createCode(Context *context, ASTNode *node)
{
    cout << "createCode" << endl;
    generateCode(context->script, node, context->staticLink, "");
};
void generateCode(Script *script, ASTNode *node, Scope *slink, string indent)
{
    indent += "\t";
    switch (node->type)
    {
    case ASTNodeType::Program:
    case ASTNodeType::Block:
        for (auto child : node->children)
            generateCode(script, child, slink, indent);
        break;
    case ASTNodeType::Function:
    {
        Atom *nameAtom = script->atoms[getAtomIndex(script, node->text)];
        //函数对象
        Script *funScript = new Script(1000);
        Function *fun = new Function(nameAtom, funScript, slink);
        generateCode(funScript, node->children[0], fun->slink, indent);
        //存储函数对象的数据
        Datum *fund = new Datum();
        fund->type = DATUM_TYPE::FUNCTION;
        fund->u.fun = fun;
        Property *funProp = new Property(fund);
        //函数符号，指向函数对象数据
        Symbol *funSym = new Symbol(SYMBOL_TYPE::PROPERTY, slink);
        funSym->entry.key = nameAtom;
        funSym->entry.value = funProp;
        pushSymbol(funSym, slink);
        //函数形参符号,倒叙的参数链
        for (int i = node->params.size() - 1; i >= 0; i--)
        {
            string param = node->params[i];
            Symbol *sym = new Symbol(SYMBOL_TYPE::ARGUMENT, fun->slink);
            sym->entry.key = funScript->atoms[getAtomIndex(funScript, param)];
            Symbol *next = funScript->args;
            sym->next = next;
            funScript->args = sym;
        }
    }
    break;
    case ASTNodeType::Call:
    {
        emit2(script, OP_TYPE::NAME, getAtomIndex(script, node->text));
        //实参入栈
        for (auto child : node->children)
            generateCode(script, child, slink, indent);
        emit2(script, OP_TYPE::CALL, node->children.size());
    }
    break;
    case ASTNodeType::Binary:
        generateCode(script, node->children[0], slink, indent);
        generateCode(script, node->children[1], slink, indent);
        if (node->text == "+")
            emit1(script, OP_TYPE::ADD);
        else if (node->text == "-")
            emit1(script, OP_TYPE::MINUS);
        else if (node->text == "*")
            emit1(script, OP_TYPE::MUTIL);
        else if (node->text == "/")
            emit1(script, OP_TYPE::DIVID);
        break;
    case ASTNodeType::Return:
        if (node->children.size())
            generateCode(script, node->children[0], slink, indent);
        emit2(script, OP_TYPE::RETURN, node->children.size() ? 1 : 0);
        break;
    case ASTNodeType::AssignmentStmt:
    case ASTNodeType::VarDeclaration:
        generateCode(script, node->children[0], slink, indent);
        emit2(script, OP_TYPE::NAME, getAtomIndex(script, node->text));
        emit1(script, OP_TYPE::ASSIGN);
        break;
    case ASTNodeType::NumberLiteral:
        emit2(script, OP_TYPE::NUMBER, getAtomIndex(script, stoi(node->text)));
        break;
    case ASTNodeType::Identifier:
        emit2(script, OP_TYPE::NAME, getAtomIndex(script, node->text));
        break;
    default:
        break;
    }
}
void emit1(Script *script, OP_TYPE type)
{
    unsigned base = script->length;
    script->code[base + 0] = (uint8_t)type;
    script->length++;
}
void emit2(Script *script, OP_TYPE type, uint8_t op1)
{
    unsigned base = script->length;
    script->code[base + 0] = (uint8_t)type;
    script->code[base + 1] = op1;
    script->length += 2;
}

File: tiny-virtual-machine-object/src/bytecode.ts
import { ASTNode, ASTNodeType, ATOM_TYPE, DATUM_TYPE, OP_TYPE, SYMOBL_TYPE, _Atom, _Context, _Datum, _Function, _Property, _Scope, _Script, _Symbol } from "./models";
/**
 * 语义解析生成字节码
 */
export function createCode(context: _Context, node: ASTNode) {
    context.script = new _Script();
    generateCode(context, context.script, node, 0, "");
}
function generateCode(context: _Context, script: _Script, node: ASTNode, offset: number = 0, indent: string) {
    indent += "\t";
    const atoms = script.atoms;
    const dataView = new DataView(script.code);
    console.log(indent, node.type, node.token?.text);
    switch (node.type) {
        case ASTNodeType.Block:
        case ASTNodeType.Program:
            node.children.forEach(child => offset = generateCode(context, script, child, offset, indent));
            break;
        case ASTNodeType.Function:
            const nameAtom = atoms[getAtomIndex(ATOM_TYPE.NAME, node.token.text)];
            const funScript = new _Script();
            const fun = new _Function(nameAtom, funScript, context.staticLink);
            fun.scope = new _Scope();
            generateCode(context, fun.script, node.children[0], 0, "");
            fun.script.args = node.args.map(arg => new _Symbol(fun.scope, SYMOBL_TYPE.ARGUMENT, { key: funScript.atoms.find(item => item.val == arg.text) }));
            fun.scope.list = fun.script.args;
            const funSym = new _Symbol(context.staticLink, SYMOBL_TYPE.PROPERTY, {
                key: nameAtom,
                value: new _Property(new _Datum(DATUM_TYPE.FUNCTION, fun))
            });
            context.staticLink.list.push(funSym);
            break;
        case ASTNodeType.Binary:
            node.children.forEach(child => offset = generateCode(context, script, child, offset, indent));
            switch (node.token.text) {
                case "+":
                    appendBuffer(OP_TYPE.ADD);
                    break;
                case "-":
                    appendBuffer(OP_TYPE.MINUS);
                    break;
                case "*":
                    appendBuffer(OP_TYPE.MUTIL);
                    break;
                case "/":
                    appendBuffer(OP_TYPE.DIVID);
                    break;
            }
            break;
        case ASTNodeType.New:
            appendBuffer(OP_TYPE.NAME, getAtomIndex(ATOM_TYPE.NAME, node.children[0].token.text));
            node.children[0].children.forEach(child => offset = generateCode(context, script, child, offset, indent));
            appendBuffer(OP_TYPE.NEW, node.children[0].children.length);
            break;
        case ASTNodeType.Call:
            appendBuffer(OP_TYPE.NAME, getAtomIndex(ATOM_TYPE.NAME, node.token.text));
            node.children.forEach(child => offset = generateCode(context, script, child, offset, indent));
            appendBuffer(OP_TYPE.CALL, node.children.length);
            break;
        case ASTNodeType.Return:
            node.children.forEach(child => offset = generateCode(context, script, child, offset, indent));
            appendBuffer(OP_TYPE.RETURN);
            break;
        case ASTNodeType.VarDeclaration:
        case ASTNodeType.AssignmentStmt:
            offset = generateCode(context, script, node.children[0], offset, indent)
            appendBuffer(OP_TYPE.NAME, getAtomIndex(ATOM_TYPE.NAME, node.token.text));
            appendBuffer(OP_TYPE.ASSIGN);
            break;
        case ASTNodeType.NumberLiteral:
            appendBuffer(OP_TYPE.NUMBER, getAtomIndex(ATOM_TYPE.NUMBER, parseInt(node.token.text)));
            break;
        case ASTNodeType.Identifier:
            appendBuffer(OP_TYPE.NAME, getAtomIndex(ATOM_TYPE.NAME, node.token.text));
            break;
    }
    if (indent.length == 1) script.code = script.code.slice(0, offset);
    return offset;
    function appendBuffer(...params: number[]) {
        params.forEach(value => {
            dataView.setUint8(offset, value);
            offset += 8;
        });
    }
    function getAtomIndex(type, val) {
        let index = atoms.findIndex(item => item.flag == type && item.val == val);
        if (index >= 0) return index;
        atoms.push(new _Atom(type, val));
        return atoms.length - 1;
    }
}